csplit/csplit.1
.\"
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)csplit.1	1.9 (gritter) 8/14/05
.TH CSPLIT 1 "8/14/05" "Heirloom Toolchest" "User Commands"
.SH NAME
csplit \- context split
.SH SYNOPSIS
\fBcsplit\fR [\fB\-s\fR] [\fB\-k\fR] [\fB\-f\fI\ prefix\fR]
[\fB\-n\fI\ digits\fR] \fIfile args\fR\ ...
.SH DESCRIPTION
.I Csplit
reads
.I file
and writes it in pieces as specified by the further operands
.IR args .
The destination file names have the form
\fBxx\fI00\fR, \fBxx\fI01, and so on by default.
.PP
The
.I args
operands can have one of the following forms:
.TP 20
\fB/\fIrexp\fB/\fR[[\fB+\fR|\fB\-\fR]\fIoffset\fR]
Write text between the current line
and the next line matching the regular expression
.I rexp
to an output file piece.
If
.I offset
is also
specified,
writing ends
.I offset
lines below the matching line for positive values,
or above the matching line for negative ones.
The text section written to the next file piece
starts at the matching line with
.IR offset applied ;
thus repeated matches with negative offsets
will repeatedly write the same text part.
With
.BR /usr/5bin/csplit ,
.I rexp
is treated as a simple regular expression,
whereas
.BR /usr/5bin/s42/csplit ,
.BR /usr/5bin/posix/csplit ,
and
.B /usr/5bin/posix2001/csplit
treat it as a basic regular expression (see
.IR ed (1).
Backslash escapes a literal
.RB ` / '
in the expression.
.TP 20
\fB%\fIrexp\fB%\fR[[\fB+\fR|\fB\-\fR]\fIoffset\fR]
Also match
.I rexp
as described above,
but skip the text between the current line
and the matching one (plus
.IR offset )
instead of writing it.
Backslash escapes
.RB ` % '.
.TP 20
\fInumber\fR
Create a piece of the file
for the text between the current line
and line
.IR number .
Line numbering starts at 1.
.TP 20
\fB{\fIrepetition\fB}\fR
Apply the previous argument additionally
.I repetition
times if it searches for a regular expression.
If it specifies a line number,
include the text between the current line
and the current line plus the line count given in the argument,
and repeat this process
.I repetition
times.
.PP
The following options can be used with
.IR csplit :
.TP
\fB\-f\fI\ prefix\fR
Use the given
.I prefix
as the first part of the file names created
(instead of
.BR xx ).
.TP
\fB\-n\fI\ digits\fR
Use
.I digits
positions for the second part of the file names created,
instead of the default 2.
This allows for the creation of more than 100 file pieces.
This option was introduced by POSIX.2.
.TP
.B \-k
When
.I csplit
detects an error
(as an argument that does not correspond to a part of the input file)
or it is interrupted from the terminal,
output files already created will be deleted.
Specifying this option causes
.I csplit
to keep these files.
.TP
.B \-s
The number of bytes written to each file piece
is normally written to standard output;
this option inhibits these statistics.
.SH EXAMPLES
Create individual files
.IR section00 ,
.I section01
\&... for each section of a
.I "troff \-ms"
document:
.RS
.sp
csplit \-k \-f section doc.ms \'/^\.[NS]H/\' {100}
.RE
.SH "ENVIRONMENT VARIABLES"
.TP
.BR LANG ", " LC_ALL
See
.IR locale (7).
.TP
.B LC_COLLATE
Affects the collation order for range expressions,
equivalence classes, and collation symbols in basic
regular expressions.
.TP
.B LC_CTYPE
Determines the mapping of bytes to characters,
the set of word boundary characters
and the composition of character classes
in basic regular expressions.
.SH "SEE ALSO"
ed(1),
split(1)


CHANGES
Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.
* cpio: Added a "-D" option to make it usable in combination with SVR4
  packaging tools. This option is not intended for general use, and is
  thus not documented.
* The package build target "make pkg" for Solaris and Open UNIX is no
  longer supported.

Release 070115
* deroff: Now recognizes the ".do" and ".xflag" requests as well as an "-xN"
  command line option to handle Heirloom nroff/troff extensions.
* ed: Substituting only a specific occurence of the pattern, as in the
  command "s/x/y/3", now works correctly if the substitution "y" is not
  equal in length to the pattern match "x" (Derek Thomas).
* ps: Use the Linux 2.6 /proc/status "SleepAVG" field to compute the value
  for the "C" column (cpu usage for scheduling).
* priocntl: The "FF", "B", and "ISO" class names are now accepted with
  the "-e" option too.
* file: UTF-8 characters in troff input files are now recognized.
* cp: The "-s" option now displays correct sizes and related statistics
  for special files with st_size == 0.
* grep: With the "-z" option, decompressor children processes are now
  properly waited for, avoiding failures when processing many files.
* su: Works with PAM on Mac OS X now (Stefan Tramm).
* ps: Handles dead processes as on other platforms now (Stefan Tramm).
* pgrep: Ignores dead processes now (Stefan Tramm).
* spell: A value of the wrong type that was passed to scanf() could lead
  to build errors on 64-bit platforms (Jeffrey H. Johnson).
* ps, pgrep, whodo: Works on FreeBSD 6 now.
* df, find, ps: Changes have been made to build on NetBSD 3 (Aleksey Cheusov).
* Changes have been made to build on OpenBSD 3.8 (Aleksey Cheusov).
* Changes have been made to build with GNU make 3.81.

Release 060122
* The tools have been ported to Mac OS X 10.4.3 (patches by Stefan Tramm).
* The tools have been ported to DragonFlyBSD 1.3.7-DEVELOPMENT.

Release 051118
* rm: The output of the question before descending to a directory with
  the "-i" option has been fixed.
* /usr/ucb/ps: The "-a" option has been fixed; it incorrectly printed a
  list of all processes since release 050411.
* getconf: /usr/ucb is now put at the end of the Heirloom directory list
  for the PATH variable.
* The "-man" macro no longer produces cut marks when used with troff.
* tr: When evaluating '[:class:]' expressions, the last position of the
  coded character set was not included; this has been fixed.
* ps: The "-R" option (resource usage format), which had been present
  for some time, has been documented.
* RPM package builds now require the new "Heirloom Development Tools"
  package from <http://heirloom.sourceforge.net/devtools.html> to be
  installed.

Release 050815
* The "getconf" utility has been added (derived from OpenSolaris code).
* The "mesg" and "pathchk" utilities have been added.
* tr: A new implementation that is based on 6th and 7th edition code is
  now used. It stores the translation tables in arrays of variable size.
  This results in better performance, under some circumstances at the
  expense of higher memory consumption.
* printf: Handle unsigned values larger than INT_MAX correctly with "%u",
  "%o", "%x", and "%X".
* printf: Handle SVID3 "position" fields in format specifications, such
  as "%7$s", that specify which argument is used next.
* pg: If a line does not fit on the screen, do not print a newline to
  force wraparound unless it is really necessary. This allows to select
  words across the wrap for copy-and-paste in an xterm.
* pg: Discard the ESC[48m sequence on a terminal that supports colors.
* ed: Do not print "expecting `w'" with the "q" command after a zero-byte
  file has been written.
* man: The tmac.an macros no longer use "@" as an escape character; its
  previous use lead to weird output with Unix nroff when an "@" appeared
  in titles or hanging labels.
* man: The "TCAT" post processor is now invoked separately for each manual
  entry. Previously, when the -t option was given with multiple entries and
  TCAT=dpost, only the first entry was actually printed.
* Some issues with building on AIX have been fixed.

Release 050713
* The 'echo' and 'test' commands have been added as standalone versions of
  those of the Heirloom Bourne shell.
* The 'false', 'kill', 'printf', 'pwd', and 'true' commands have been added.
* ed: Copying a larger number of lines with the 't' command could cause a
  segmentation fault because some pointers were not updated after calling
  realloc().
* ed: Do not warn about an 'input error' anymore when reading a file that
  lacks the terminating newline.
* ed: Disallow the 'u' command if no change has been made yet after a file
  specified on the command line has been read.

csplit/csplit.c
/*
 * csplit - context split
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, May 2003.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * This software is provided 'as-is', without any express or implied
 * warranty. In no event will the authors be held liable for any damages
 * arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute
 * it freely, subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not
 *    claim that you wrote the original software. If you use this software
 *    in a product, an acknowledgment in the product documentation would be
 *    appreciated but is not required.
 *
 * 2. Altered source versions must be plainly marked as such, and must not be
 *    misrepresented as being the original software.
 *
 * 3. This notice may not be removed or altered from any source distribution.
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#if defined (SU3)
static const char sccsid[] USED = "@(#)csplit_su3.sl	1.10 (gritter) 5/29/05";
#elif defined (SUS)
static const char sccsid[] USED = "@(#)csplit_sus.sl	1.10 (gritter) 5/29/05";
#else
static const char sccsid[] USED = "@(#)csplit.sl	1.10 (gritter) 5/29/05";
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <unistd.h>
#include <stdarg.h>
#include <math.h>
#include <locale.h>
#include <signal.h>
#include "sigset.h"
#include "atoll.h"

#if defined (__GLIBC__) 
#if defined (_IO_getc_unlocked)
#undef	getc
#define	getc(f)		_IO_getc_unlocked(f)
#endif
#if defined (_IO_putc_unlocked)
#undef	putc
#define	putc(c, f)	_IO_putc_unlocked(c, f)
#endif
#endif

#if defined (SUS) || defined (SU3)
#include <regex.h>
#else
#include <regexpr.h>
#endif

#include <iblok.h>

static struct	arg {
	long long	a_no;
	long long	a_nx;
	long long	a_ro;
	long long	a_once;
	const char	*a_op;
	const char	*a_rp;
#if defined (SUS) || defined (SU3)
	regex_t		*a_re;
#else
	char		*a_re;
#endif
} **args;

static const char	*progname;
static const char	*prefix = "xx";
static const char	*curarg;
static int		filec;
static int		suffixlength = 2;
static int		status;
static int		kflag;
static int		sflag;

static void	usage(void);
static void	msg(int, const char *, ...);
static void	scan(char *);
static void	csplit(const char *);
static int 	match(const char *, long long, long long *, int *);
static FILE	*nextfile(void);
static const char	*makename(int, int);
static void	delfiles(void);
static void	onint(int);
static struct iblok	*copytemp(struct iblok *);
static void	*smalloc(size_t);
static void	*srealloc(void *, size_t);
static void	*scalloc(size_t, size_t);

int
main(int argc, char **argv)
{
	int	i, illegal = 0;
	char	*fn;

	progname = basename(argv[0]);
#if defined (SUS) || defined (SU3)
	setlocale(LC_COLLATE, "");
#endif
	setlocale(LC_CTYPE, "");
	while ((i = getopt(argc, argv, "ksf:n:")) != EOF) {
		switch (i) {
		case 'k':
			kflag = 1;
			break;
		case 's':
			sflag = 1;
			break;
		case 'f':
			prefix = optarg;
			break;
		case 'n':
			suffixlength = atoi(optarg);
			break;
		default:
			illegal++;
		}
	}
	if (illegal || optind + 2 > argc)
		usage();
	fn = argv[optind++];
	args = scalloc(argc - optind + 1, sizeof *args);
	for (i = 0; i + optind < argc; i++)
		scan(argv[i+optind]);
	if (sigset(SIGINT, SIG_IGN) != SIG_IGN)
		sigset(SIGINT, onint);
	csplit(fn);
	return status;
}

static void
usage(void)
{
	fprintf(stderr, "%s: Usage: %s [-s] [-k] [-f prefix] file args ...\n",
		progname, progname);
	exit(2);
}

static void
msg(int term, const char *fmt, ...)
{
	va_list	ap;

	fprintf(stderr, "%s: ", progname);
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	putc('\n', stderr);
	delfiles();
	if (term)
		exit(1);
	else
		status |= 1;
}

static void
scan(char *s)
{
	static int	ncur = -1;
	long long	c;
	char	*sp, *x;
#if defined (SUS) || defined (SU3)
	int	reflags;
#endif

	if (*s != '{') {
		args[++ncur] = scalloc(1, sizeof *args[ncur]);
		args[ncur]->a_op = s;
	}
	switch (*s) {
	case '/':
	case '%':
		for (sp = &s[1]; *sp; sp++) {
			if (sp[0] == '\\' && sp[1])
				sp++;
			else if (*sp == *s)
				break;
		}
		if (*sp == '\0')
			msg(1, "%s: missing delimiter", s);
		*sp = '\0';
		args[ncur]->a_nx = strtoll(&sp[1], &x, 10);
		if (*x != '\0')
			msg(1, "%s: illegal offset", s);
#if defined (SUS) || defined (SU3)
		args[ncur]->a_re = smalloc(sizeof *args[ncur]->a_re);
		reflags = REG_ANGLES|REG_NOSUB;
#if defined (SU3)
		reflags |= REG_AVOIDNULL;
#endif	/* SU3 */
		if (regcomp(args[ncur]->a_re, &s[1], reflags) != 0)
#else	/* !SUS, !SU3 */
		if ((args[ncur]->a_re = compile(&s[1], 0, 0)) == 0)
#endif	/* !SUS, !SU3 */
			msg(1, "%s: Illegal Regular Expression", s);
		break;
	default:
		args[ncur]->a_nx = args[ncur]->a_no = strtoll(s, &x, 10);
		if (*x != '\0')
			msg(1, "%s: bad line number", s);
		break;
	case '{':
		c = strtoll(&s[1], &x, 10);
		if (x > &s[1] && x[0] != '}')
			msg(1, "%s: missing '}'", s);
		if (s[1] == '-' || x[0] != '}')
			msg(1, "Illegal repeat count: %s", s);
		if (ncur < 0 || args[ncur] == 0 || args[ncur]->a_rp)
			msg(1, "No operation for %s", s);
		args[ncur]->a_ro = c;
		args[ncur]->a_rp = s;
	}
}

static void
csplit(const char *fn)
{
	struct iblok	*ip;
	int	skip = 0, osk = 0, gotcha;
	FILE	*op = NULL;
	char	*line = 0, c = 0;
	size_t	linesize = 0, linelen;
	long long	lineno = 0, oln = 1, xln = 1,
			bytes = 0, noffs = 0, on, brks = 0;

	if ((ip = fn[0] == '-' && fn[1] == '\0' ? ib_alloc(0, 0) :
			ib_open(fn, 0)) == NULL)
		msg(1, "Cannot open %s", fn);
	if (ib_seek(ip, 0, SEEK_CUR) != 0)
		ip = copytemp(ip);
	do {
		if ((linelen=ib_getlin(ip, &line, &linesize, srealloc)) != 0) {
			lineno++;
			if ((c = line[linelen-1]) == '\n')
				line[linelen-1] = '\0';
		} else
			lineno = -1;
		while (on = noffs, osk = skip,
		    (gotcha=match(line, lineno, &noffs, &skip)) != 0) {
			if (on) {
				if (osk == 0)
					op = nextfile();
				ib_seek(ip, brks, SEEK_SET);
				on = xln - on + (lineno <= 0);
				lineno = oln - 1;
				if (lineno > on)
					msg(1, "%s - out of range", curarg);
				while ((linelen = ib_getlin(ip, &line,
							&linesize,
							srealloc)) != 0 &&
						lineno++ < on) {
					if (osk == 0) {
						fwrite(line, sizeof *line,
								linelen, op);
						bytes += linelen;
					}
				}
				if (linelen != 0 && (c=line[linelen-1]) == '\n')
					line[linelen-1] = '\0';
			}
			if (lineno==1 && op==NULL && *args[0]->a_op != '%' &&
					(args[0]->a_re==0 || args[0]->a_nx>=0))
				op = nextfile();
			if (op) {
				if (!sflag)
					printf("%lld\n", bytes);
				bytes = 0;
				fclose(op);
			}
			if (lineno <= 0)
				break;
			if (!skip) {
				if (noffs) {
					op = NULL;
					brks = ib_seek(ip, 0, SEEK_CUR) -
						linelen;
					oln = lineno;
				} else
					op = nextfile();
			} else
				op = NULL;
			if (gotcha < 2)
				break;
		}
		if (!skip) {
			if (lineno == 1 && op == NULL && noffs == 0)
				op = nextfile();
			if (op && linelen != 0) {
				if (c == '\n')
					line[linelen-1] = '\n';
				fwrite(line, sizeof *line, linelen, op);
				bytes += linelen;
			}
		}
		xln = lineno;
	} while (linelen != 0);
}

static int
match(const char *line, long long lineno, long long *noffp, int *skip)
{
	static int	ncur = -1;

	if (ncur == -1) {
		ncur = 0;
		curarg = args[ncur]->a_op;
		if (args[ncur]->a_re)
			*skip = *args[ncur]->a_op == '%';
		if (args[ncur]->a_re && args[ncur]->a_nx < 0)
			*noffp = -args[ncur]->a_nx;
		else
			*noffp = 0;
	}
	if (args[ncur] == NULL)
		return lineno <= 0;
	if (args[ncur]->a_once == lineno)
		return 0;
	args[ncur]->a_once = 0;
	*skip = *args[ncur]->a_op == '%';
	if (args[ncur]->a_re != NULL && lineno > 0) {
		if (args[ncur]->a_no) {
			if (--args[ncur]->a_no > 0)
				return 0;
		} else {
#if defined (SUS) || defined (SU3)
			if (regexec(args[ncur]->a_re, line, 0, NULL, 0) != 0)
#else
			if (step(line, args[ncur]->a_re) == 0)
#endif
				return 0;
			if (args[ncur]->a_nx > 0) {
				args[ncur]->a_no = args[ncur]->a_nx;
				return 0;
			}
			if (args[ncur]->a_nx < 0 &&
					args[ncur]->a_nx+lineno < 0)
				msg(1, "%s - out of range", curarg);
		}
	} else {
		if (lineno > 0 && lineno < args[ncur]->a_no)
			return 0;
		else if (lineno < 0 && args[ncur]->a_nx < 0)
			/*EMPTY*/;
		else if (lineno < 0 || lineno > args[ncur]->a_no)
			msg(kflag==0||lineno>0, "%s - out of range", curarg);
	}
	if (args[ncur]->a_re && args[ncur]->a_nx >= 0)
		args[ncur]->a_once = lineno;
	if (args[ncur]->a_ro) {
		args[ncur]->a_ro--;
		curarg = args[ncur]->a_rp;
		if (args[ncur]->a_re == 0)
			args[ncur]->a_no += args[ncur]->a_nx;
	} else {
		if (args[++ncur]) {
			curarg = args[ncur]->a_op;
			if (args[ncur]->a_re)
				*skip = *args[ncur]->a_op == '%';
			else
				*skip = 0;
			if (args[ncur]->a_re && args[ncur]->a_nx < 0)
				*noffp = -args[ncur]->a_nx;
			else
				*noffp = 0;
		} else {
			*noffp = 0;
			*skip = 0;
		}
	}
	return 1 + ((lineno==1&&ncur==0) || args[ncur]==0 ||
			args[ncur]->a_re==0 || args[ncur]->a_nx < 0);
}

static FILE *
nextfile(void)
{
	const char	*name;
	FILE	*fp;

	name = makename(filec++, 1);
	if ((fp = fopen(name, "w")) == NULL)
		msg(1, "Cannot create %s", name);
	return fp;
}

static const char *
makename(int n, int prnt)
{
	static char		*name, *sufp;
	int	c;
	const char	*cp;
	char	*sp;

	if (name == 0) {
		name = smalloc(strlen(prefix)+suffixlength+1);
		sufp = name;
		for (cp = prefix; *cp; cp++)
			*sufp++ = *cp;
	}
	c = n;
	sp = &sufp[suffixlength];
	*sp = '\0';
	while (--sp >= sufp) {
		*sp = (c % 10) + '0';
		c /= 10;
	}
	if (c) {
		if (prnt)
			msg(1, "%0.0f file limit reached at arg %s",
				pow(10, suffixlength), curarg);
		return 0;
	}
	return name;
}

static void
delfiles(void)
{
	const char	*np;
	int	i;

	if (!kflag)
		for (i = 0; i < filec && (np = makename(i, 0)); i++)
			unlink(np);
}

static struct iblok *
copytemp(struct iblok *ip)
{
	char	tfname[] = "/tmp/csplitXXXXXX";
	char	buf[4096];
	struct iblok	*op = NULL;
	int	fd;
	ssize_t	rd;

	if ((fd = mkstemp(tfname)) < 0 || unlink(tfname) < 0)
		goto err;
	while ((rd = read(ip->ib_fd, buf, sizeof buf)) > 0)
		if (write(fd, buf, rd) != rd)
			goto err;
	if (ip->ib_fd)
		ib_close(ip);
	else
		ib_free(ip);
	if (lseek(fd, 0, SEEK_SET) != 0 || (op = ib_alloc(fd, 0)) == NULL)
	err:	msg(1, "Bad write to temporary file");
	return op;
}

/*ARGSUSED*/
static void
onint(int signo)
{
	msg(1, "Interrupt - program aborted at arg '%s'", curarg);
}

static void *
smalloc(size_t size)
{
	return srealloc(0, size);
}

static void *
srealloc(void *op, size_t size)
{
	void	*np;

	if ((np = realloc(op, size)) == NULL) {
		write(2, "no memory\n", 10);
		_exit(077);
	}
	return np;
}

static void *
scalloc(size_t nmemb, size_t size)
{
	void	*vp;

	if ((vp = calloc(nmemb, size)) == NULL) {
		write(2, "no memory\n", 10);
		_exit(077);
	}
	return vp;
}
